Level 1

DenisIlyushin opened the pull request.

DenisIlyushin committed:
feat: подготовка проекта

DenisIlyushin committed:
feat: структура проекта

DenisIlyushin committed:
feat: добавить модели

DenisIlyushin committed:
feat: добавить обработку ошибок из предыдущего проекта

DenisIlyushin committed:
feat: добавить мидлвары

DenisIlyushin committed:
feat: работа над роутами

DenisIlyushin committed:
feat: отладить бекенд локально

DenisIlyushin committed:
feat: отладить бекенд локально

DenisIlyushin committed:
feat: codestyle

DenisIlyushin committed:
feat: отладка работы на сервере

DenisIlyushin committed:
feat: создать README.md

DenisIlyushin committed:
fix: checklist

DenisIlyushin committed:
fix: checklist

DenisIlyushin committed:
fix: eslint

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [ ] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [ ] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
  **Присутствуют блокировки вида `eslint-disable`.**
- [ ] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
**Ошибок не должно быть без блокировок.**
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [ ] `PATCH /users/me` — обновляет информацию о пользователе;
  **Запрос не работает.**  
![2023-09-14_06-01-28](https://github.com/DenisIlyushin/movies-explorer-api/assets/86304001/c98fd54e-fe36-4a14-b856-c807e895a5c2)
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [ ] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
**Не защищен роут `'URI не найден'`.**
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  **1. Неверный набор полей в запросе редактирования пользователя.**
  **2. Поля в запросе редактирования пользователя должны быть обязательными.**
  **3. Поле `name` в запросе регистрации должно быть обязательным.**
  - [ ] Данные валидируются перед добавлением в базу. 
  **Поле `name` на уровне схемы пользователя должно быть обязательным.**
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
**Валидация описана в роутах.**
- [ ] <!--2.14--> Настроено логгирование:
  - [ ] запросы и ответы записываются в файл request.log;
  **Логгер запросов подключен после лимитера. Запросы, превышающие лимит, не попадают в лог.**
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.85--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co или API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Сообщения прописаны вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 55

DenisIlyushin committed:
fix: исправления после Rev1

1. Удалены директивы eslint-disable в коде, даже для блоков кода, работающих только в режиме разработки
2. Все роуты, кроме `/signin` и `/signup`, защищены авторизацией
3. Исправлена работа URI `PATCH /users/me`
- Исправлен набор полей в запросе редактирования пользователя
- Поле `name` в запросе редактирования пользователя обязательно
- Поле `name` в модели пользователя обязательно
4. Валидация описана в отдельном модуле ./utils/routesValidators.js
5. Логгер запросов теперь подключен до rateLimiter
6. Сообщения ошибок выненсены в отдельные модули в ./errors/messages/...

DenisIlyushin commented:
Сначала думал в списке галочки ставить, что поправил, но понял, что это не самая светлая идея.

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  **Поле `name` в запросе регистрации должно быть обязательным.**
  - [x] Данные валидируются перед добавлением в базу. 
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.85--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co или API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Присутствуют сообщения, прописанные вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 92
